shingled_disassembly/conservative_disasm.ml

```diff
@@ -0,0 +1,53 @@
+open Core.Std
+open Or_error
+open Bap.Std
+open Bap_plugins.Std
+module Dis = Disasm_expert.Basic
+
+exception Create_mem of Error.t
+module Lifted_image=Lifted_image_piqi
+
+let lifted_of_bil instruction length offset =
+  { Lifted_image.Lifted_instruction.instruction=Bil_piqi.stmts_to_piqi instruction; length; offset }
+
+let jump global_mem base_addr max_addr lift_insn state mem insn (current_elem, bil_disasm) =
```

Comment:
Thanks, I will make all requested changes that are direct and simple. I
will use the alternative match statement, the to_bits, investigate the -1,
and return:ident, drop core for core_kernel, and coalesce parameters to
record.

In terms of trade offs for semantics, I'm not sure what consumers should
want. A conservative byte offset disassembler is different in purpose than
a shingled disassembler. Shingled disassembly *will* need to know if
anything is a valid instruction and it's length at least, but we personally
don't use conservative disassembly for anything other than executing over
semantics in order to build an interpreter. If we can't interpret things
like fence instructions, we can only limit scope. Maybe just having the
shingled disassembler somehow parameterize the conservative offset
disassembler with just the bare bones logic needed to provide confidence
information about instructions and non-instruction sequences would provide
the best code reuse. Possibly it could be combined with phoenix to provide
information about function entrances and exits, and these two could
influence the confidences of one another.

So, if I do all of the above, but just provide a no-op lambda and abstract
over some functors so that the different methods can be exposed at an
interface level with regards to shingled and conservative disassembly,
would you accept that?

On Wed, Aug 19, 2015 at 2:10 PM, Ivan Gotovchits <notifications@github.com>
wrote:

> In shingled_disassembly/conservative_disasm.ml
> <https://github.com/BinaryAnalysisPlatform/bap-plugins/pull/3#discussion_r37447691>
> :
>
> > @@ -0,0 +1,53 @@
> > +open Core.Std
> > +open Or_error
> > +open Bap.Std
> > +open Bap_plugins.Std
> > +module Dis = Disasm_expert.Basic
> > +
> > +exception Create_mem of Error.t
> > +module Lifted_image=Lifted_image_piqi
> > +
> > +let lifted_of_bil instruction length offset =
> > +  { Lifted_image.Lifted_instruction.instruction=Bil_piqi.stmts_to_piqi instruction; length; offset }
> > +
> > +let jump global_mem base_addr max_addr lift_insn state mem insn (current_elem, bil_disasm) =
>
> when you have so many parameters, it is better to coalesce it into a
> record type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BinaryAnalysisPlatform/bap-plugins/pull/3/files#r37447691>
> .
>